OSDOCS-10472:nodes peer ANP (#75886)
openshift/openshift-docs: modules/nw-anp-nodes-peer-concept.adoc

// Module included in the following assemblies:
//
// * networking/openshift_network_security/AdminNetworkPolicy/nw-nodes-peer.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-egress_{context}"]
= Egress traffic control for AdminNetworkPolicy and BaselineAdminNetworkPolicy

As a cluster administrator you can use `AdminNetworkPolicy` (ANP) resource in conjunction with `BaselineAdminNetworkPolicy` (BANP) resource to secure and limit egress traffic. Using these namespace scoped policies together you can do the following:

* Create egress traffic control for cluster nodes.

* Apply egress traffic control towards Kubernetes API servers.

* Implement egress traffic control towards external destinations.

//For example, as a cluster administrator, you can enforce security for northbound traffic by restricting traffic from specific cluster workloads to all or specific destinations outside the cluster. Or, you can setup policies that grant access to your cluster-external DNS server even if namespace admins create `NetworkPolicy` resources that block cluster-external egress.

[id="egress-traffic-control-towards-cluster-nodes_{context}"]
== Using egress traffic control for ANP and BANP

> Using the `matchLabels` and `matchExpressions` fields, cluster administrators can select nodes in a cluster based on `nodes` and `networks` labels and apply egress traffic policy to these nodes. The following are benefits to using label selectors when blocking access from cluster workloads to specific ports on cluster nodes:

Reviewer: matchLabels and/or matchExpressions are for nodes. networks have the multiple CIDRs

Author: I tried to separate the distinctions more. Let me know if I've got something wrong.

* You do not have to block access to those ports on external hosts.

* You do not have to manually list the IP address of every node.

* You do not have to change the policy when new nodes are added to the cluster.

The following example applies `matchExpression` key value pairs to deny egress traffic from tenants who do not have a `restricted`, `confidential`, or `internal` level of security. In the example, the `Deny` action is applied using the `nodes.matchLabels` field so that security clearance is needed to access control-plane nodes at `443` and `6443` ports in the cluster.

.Example of ANP `Deny` egress to control-plane
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: egress-security-deny
spec:
  priority: 55
  subject:
    namespaces:
      matchExpressions:
        - {key: security, operator: notIn, values: [restricted, confidential, internal]} <1>
  egress:

>   - name: "deny-all-egress-to-kapi-server"

Reviewer 2: We do not put " " around name, please remove " "

>     action: "Deny"

Reviewer 2: We do not put " " around action, please remove " "

    to:
    - nodes: <2>
        matchLabels:
          node-role.kubernetes.io/control-plane:
    ports:
      - portNumber:
          protocol: TCP
          port: 443
      - portNumber:
          protocol: TCP
          port: 6443
----

> <1> Specifies key/value pairs for `matchExpressions` field.

Reviewer: You may want to use matchExpression in the example as you are describing it.
<pre>
- nodes:
    matchExpressions:
     - key: node-role.kubernetes.io/control-plane
       operator: Exists
</pre>

<2> Specifies a node or set of nodes in the cluster using the `matchLabels`. It will select all nodes if the label is present, but left empty.
====

The following example uses `networks` field and combines ANP and BANP polices to isolate egress traffic to a range of specified CIDR.

.Example of ANP and BANP using `networks` field and CIDR
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: network-as-egress-peer
spec:
  priority: 70
  subject:
    namespaces: {}
  egress:

>   - name: "deny-egress-to-external-dns-servers"

Reviewer 2: We do not put " " around name, please remove " "

>     action: "Deny"

Reviewer 2: We do not put " " around name, please remove " "

    to:
    - networks: <1>
      - 194.0.2.0/24
      - 205.0.113.15/32
      - 199.51.100.10/32
    ports:
      - portNumber:
          protocol: UDP
          port: 53

>   - name: "allow-all-egress-to-intranet"

Reviewer 2: We do not put " " around name, please remove " "

>     action: "Allow"

Reviewer 2: We do not put " " around action, please remove " "

    to:
    - networks:
      - 192.0.2.0/24
      - 203.0.113.0/24
      - 198.51.100.0/24

>   - name: "allow-all-intra-cluster-traffic"

Reviewer 2: We do not put " " around name, please remove " "

>     action: "Allow"

Reviewer 2: We do not put " " around action, please remove " "

    to:
    - networks: <2>
      - <POD_CIDR>
      - <NODE_CIDR>
      - <SERVICE_CIDR>

>   - name: "pass-all-egress-to-internet"

Reviewer 2: We do not put " " around name, please remove " "

>     action: "Pass"

Reviewer 2: We do not put " " around action, please remove " "

    to:
    - networks:
      - 0.0.0.0/0
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces: {}
  egress:

>   - name: "deny-all-egress-to-internet"

Reviewer 2: We do not put " " around name, please remove " "

>     action: "Deny"

Reviewer 2: We do not put " " around action, please remove " "

// * networking/network_security/AdminNetworkPolicy/ovn-k-egress-nodes-networks-peer.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-egress_{context}"]
= Egress traffic control for AdminNetworkPolicy and BaselineAdminNetworkPolicy

As a cluster administrator you can use the `AdminNetworkPolicy` (ANP) resource in conjunction with the `BaselineAdminNetworkPolicy` (BANP) resource to secure and limit egress traffic. Using these cluster scoped policies together you can do the following:

* Create egress traffic control for cluster nodes.

* Apply egress traffic control towards Kubernetes API servers.

* Implement egress traffic control towards external destinations.

[id="egress-traffic-control-towards-cluster-nodes_{context}"]
== Using egress traffic control for ANP and BANP

Using the `matchLabels` and `matchExpressions` fields, cluster administrators can select `nodes` peer and apply egress traffic policy to those nodes.

Cluster administrators can also use CIDR ranges for `networks` peer and apply egress traffic policy to those nodes. The following are benefits to using `nodes` and `networks` peer when blocking access from cluster workloads to specific ports on cluster nodes:

> * You do not have to block access to those ports on external hosts.

Reviewer: each external host

> * You do not have to block access to those ports on each external hosts.

Reviewer 3: I don't follow this point, please help me understand

> Using the `matchLabels` and `matchExpressions` fields, cluster administrators can select `nodes` peers and apply egress traffic policy to those nodes. Cluster administrators can also use CIDR ranges for `networks` peers and apply egress traffic policy to those nodes. The following are benefits to using `nodes` and `networks` peers when blocking access from cluster workloads to specific ports on cluster nodes:

Reviewer: Also mention that node and network peers are egress only

* You do not have to manually list the IP address of every node.

* You do not have to change the policy when new nodes are added to the cluster.

> The following example allows egress traffic to an API server on port `443` and `6443` by any of the namespaces with a `restricted`, `confidential`, or `internal` level of security and in the `dept: engr` tenancy.

Reviewer 3: please check what's running at 6443 on the cluster, I always forget if its openshift-apiserver or kube-api-server and be exact on that piece. Not "AN API" server

Reviewer: The endpoints of kube api server listen on 6443 and svc itself listens on 443.
namespace is openshift-kube-apiserver , service is apiserver

.Example of ANP `Allow` egress using `nodes` peer
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: egress-security-allow
spec:
  priority: 55
  subject:
    namespaces:
      matchExpressions: <1>
        - {key: security, operator: In, values: [restricted, confidential, internal]}
  egress:
  - name: "allow-to-api-server"
    action: "Allow"
    to:
    - nodes:
        matchExpressions: <2>

>         - key: node-role.kubernetes.io/worker

Reviewer 3: the API server does not run on worker nodes its a control plane component  -> have you tested your yamls on a live cluster and ensured the rules being written here is correctly working as expected?

>   - name: "allow-to-api-server"

Reviewer 2: please remove "" around the name.

>     action: "Allow"

Reviewer 2: please remove "" around the action

          operator: Exists
    - pods:
        namespaceSelector:
          matchLabels:
            dept: engr <3>
        podSelector: {}
    ports:
      - portNumber:
          protocol: TCP
          port: 443
      - portNumber:
          protocol: TCP

>           port: 6443

Reviewer 3: please have a deny rule to all worker nodes otherwise the allow makes less sense. currently w or w/o your ANP everything will work the same right?

Author: Arti's new example has a deny rule for them.

----
<1> Specifies key/value pairs for `matchExpressions` field.
<2> Specifies a node or set of nodes in the cluster using the `matchExpressions` field. It will select all nodes if the label is present, but left empty.
<3> Specifies all the pods labeled with `dept: engr`.
====

The following example uses `networks` peer and combines ANP and BANP polices to isolate egress traffic to a range of specified CIDR.

.Example of ANP and BANP using `networks` peers
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: network-as-egress-peer
spec:
  priority: 70
  subject:
    namespaces: {}
  egress:
  - name: "deny-egress-to-external-dns-servers"
    action: "Deny"
    to:
    - networks: <1>
      - 194.0.2.0/24
      - 205.0.113.15/32
      - 199.51.100.10/32
    ports:
      - portNumber:
          protocol: UDP
          port: 53
  - name: "allow-all-egress-to-intranet"
    action: "Allow"
    to:
    - networks:

>       - 192.0.2.0/24

Author: @asood-rh

      - 203.0.113.0/24
      - 198.51.100.0/24
  - name: "allow-all-intra-cluster-traffic"
    action: "Allow"
    to:
    - networks: <2>

>       - <POD_CIDR>

Author: I saw this in the u/s docs, but wasn't sure how to update the policy. @asood-rh can you help update this please?

      - <NODE_CIDR>
      - <SERVICE_CIDR>
  - name: "pass-all-egress-to-internet"
    action: "Pass"
    to:
    - networks:
      - 0.0.0.0/0
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces: {}
  egress:
  - name: "deny-all-egress-to-internet"
    action: "Deny"
    to:
    - networks: <3>
      - 0.0.0.0/0
----
<1> Use `networks` to specify a range of CIDR networks outside of the cluster.

> <2> Specifies the ranges for the intranet traffic from the your <POD_CIDR>, <NODE_CIDR>, <SERVICE_CIDR> resources.

Reviewer: intra cluster traffic instead of intranet traffic

<3> Specifies a strong `Deny` egress policy to the internet by setting `networks` values to - 0.0.0.0/0.
====

Collectively the "Example of ANP and BANP using `networks` peers enforces the following egress policy:

* All pods cannot talk to company's intranet DNS servers.

* All pods can talk to rest of the company's intranet.

* All pods can talk to other pods, nodes, services.

* All pods cannot talk to internet. Combining the last ANP `Pass` rule and the strong BANP `Deny` a guardrail policy is created that secures traffic in the cluster.

[id="combined-nodes-peer-networks-peer-anp"]
== Using nodes peer and networks peer together

> Cluster Administrators can combine `nodes` and `networks` peer in your ANP and BANP policies to create multi-tenant isolation that is cluster scoped.

Reviewer 3: hmm using both policies example is fine, but the concept "multi-tenant" isolation does not fit in is what I meant since the policies are not doing that. sorry if I was not clear.

> .Example of combined `nodes` and `networks` peer

Author: @asood-rh

> .Example of combined `nodes` and `networks` peer

Reviewer: @JoeAldinger Use latest policy as you have test around All pods in port line number 168, 170, 172 and 174

> .Example of combined `nodes` and `networks` peer

Author: done

> .Example of combined `nodes` and `networks` peer

Reviewer 3: do we want to change the title or leave it?

[%collapsible]
====
[source,yaml]
----

> apiVersion: policy.networking.k8s.io/v1alpha1

Author: Its been changed 3 or 4 times now.

> apiVersion: policy.networking.k8s.io/v1alpha1

Author: @asood-rh I believe Arti checked this.

kind: AdminNetworkPolicy

> metadata:

Reviewer 3: yes

> metadata:

Author: I'm fairly certain we got this worked out now.

// Module included in the following assemblies:
//
// * networking/network_security/AdminNetworkPolicy/ovn-k-egress-nodes-networks-peer.adoc

:_mod-docs-content-type: CONCEPT
[id="nw-anp-egress_{context}"]
= Northbound traffic controls for AdminNetworkPolicy and BaselineAdminNetworkPolicy

In addition to supporting east-west traffic controls, ANP and BANP also allow administrators to control their northbound traffic leaving the cluster or traffic leaving the node to other nodes in the cluster. End-users can do the following:

> * Implement egress traffic control towards cluster nodes using `nodes` egress peer.
> 
> * Implement egress traffic control towards Kubernetes API servers using `nodes` or `networks` egress peers.
> 
> * Implement egress traffic control towards external destinations outside the cluster using `networks` peer.

Reviewer 4: ```suggestion
* Implement egress traffic control towards cluster nodes using `nodes` egress peer

* Implement egress traffic control towards Kubernetes API servers using `nodes` or `networks` egress peers

* Implement egress traffic control towards destinations outside the cluster using `networks` peer
```
I believe periods are not needed for incomplete sentences, but feel free to leave them if I am wrong

[NOTE]
====

> For ANP and BANP, `nodes` and `networks` peer is not supported as ingress peer.

Reviewer 3: ```suggestion
For ANP and BANP, `nodes` and `networks` peers can be specified from the egress rules only.
```

====

[id="egress-traffic-control-towards-cluster-nodes_{context}"]
== Using nodes peer to control egress traffic to cluster nodes

> Using the `nodes` peer administrators can control egress traffic from pods to nodes in the cluster. A benefit of this is that you do not have to change the policy when nodes are added or deleted to the cluster.

Reviewer 3: ```suggestion
Using the `nodes` peer administrators can control egress traffic from pods to nodes in the cluster. A benefit of this is that you do not have to change the policy when nodes are added to or deleted from the cluster.
```

> The following example allows egress traffic to the Kubernetes API server on port `6443` by any of the namespaces with a `restricted`, `confidential`, or `internal` level of security and in the `dept: engr` tenancy.

Reviewer 4: ```suggestion
The following example allows egress traffic to the Kubernetes API server on port `6443` by any of the namespaces with a `restricted`, `confidential`, or `internal` level of security and with the `dept: engr` tenancy label.
```

> The following example allows egress traffic to the Kubernetes API server on port `6443` by any of the namespaces with a `restricted`, `confidential`, or `internal` level of security using the node selector peer.

Reviewer 3: ```suggestion
The following example allows egress traffic to the Kubernetes API server on port `6443` by any of the namespaces with a `restricted`, `confidential`, or `internal` level of security using the node selector peer. It also denies traffic to all worker nodes in your cluster from any of the namespaces with a `restricted`, `confidential`, or `internal` level of security.
```

.Example of ANP `Allow` egress using `nodes` peer
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: egress-security-allow
spec:
  egress:
  - action: Deny
    to:
    - nodes:
        matchExpressions:
        - key: node-role.kubernetes.io/worker
          operator: Exists
  - action: Allow

>     name: allow-to-kubernetes-api-server

Reviewer 3: ```suggestion
    name: allow-to-kubernetes-api-server-and-engr-dept-pods
```

    ports:
    - portNumber:
        port: 6443
        protocol: TCP

> 

Reviewer 3: why this gap?

Author: Not sure, weird. gone now

        - key: node-role.kubernetes.io/control-plane <1>
          operator: Exists
    - pods:
        namespaceSelector:
          matchLabels:
            dept: engr <2>
        podSelector: {}
  priority: 55
  subject: <3>
    namespaces:
      matchExpressions:
      - key: security
        operator: In
        values: <4>
        - restricted
        - confidential
        - internal
----

> <1> Specifies a node or set of nodes in the cluster using the `matchExpressions` field. It will select all nodes if the label is present, but left empty.

Author: Yeah I think {} that was what I was getting at.

> <1> Specifies a node or set of nodes in the cluster using the `matchExpressions` field. It will select all nodes if the label is present, but left empty.

Reviewer 3: so that means empty selector (both matchLabels and matchExpressions wise) is what you want; empty selector selects all nodes in the cluster

> <2> Specifies all the pods labeled with `dept: engr`.

Reviewer 3: like said above move it to "pods" and I think we don't need 4 ? but I'll leave it to you

> <3> Specifies the subject of the ANP which includes any namespaces that match the labels used by the network policy. In the example, any of the namespaces with `restricted`, `confidential`, or `internal` level of security.

Reviewer 4: ```suggestion
<3> Specifies the subject of the ANP which includes any namespaces that match the labels used by the network policy. The example matches any namespace with the `restricted`, `confidential`, or `internal` level of security.
```
Reworded this to avoid run-on sentence

<4> Specifies key/value pairs for `matchExpressions` field.
====

[id="egress-traffic-control-networks-peer-external-destinations_{context}"]
== Using networks peer to control egress traffic towards external destinations

> Cluster administrators can also use CIDR ranges for `networks` peer and apply egress traffic policy to those nodes. The following are benefits to using `nodes` and `networks` peer when blocking access from cluster workloads to specific ports on cluster nodes:

Reviewer 4: ```suggestion
Cluster administrators can also use CIDR ranges for `networks` peers and apply an egress traffic policy to those nodes. The following are benefits to using `nodes` and `networks` peers when blocking access from cluster workloads to specific ports on cluster nodes:
```
Some grammar nits. Plz avoid if I misunderstand the part of speech of "peer" here

> Cluster administrators can use CIDR ranges for `networks` peer and apply an egress traffic policy to those nodes.

Reviewer 3: to what nodes? (those stands for...)?

Author: nodes they specify in the policy?

Reviewer: @JoeAldinger It should say egress traffic policy to traffic leaving from node for destination configured at IP address that is within CIDR range specified with networks.

Author: updated

> * You do not have to block access to those ports on each external hosts.

Reviewer 4: ```suggestion
* You do not have to block access to those ports on each external host.
```

> Cluster administrators can use CIDR ranges for `networks` peer and apply an egress traffic policy to traffic leaving from a node and going to a destination configured at the IP address that is within the CIDR range specified with `networks` field.

Reviewer 3: ```suggestion
Cluster administrators can use CIDR ranges in `networks` peer and apply a policy to control egress traffic leaving from pods and going to a destination configured at the IP address that is within the CIDR range specified with `networks` field.
```

> The following example uses `networks` peer and combines ANP and BANP policies to isolate egress traffic to a range of specified CIDR ranges.

Reviewer 3: ```suggestion
The following example uses `networks` peer and combines ANP and BANP policies to restrict egress traffic.
```

> [IMPORTANT]

Reviewer 3: well done!

> The following example uses `networks` peer and combines ANP and BANP polices to isolate egress traffic to a range of specified CIDR.

Reviewer 4: ```suggestion
The following example uses `networks` peers and combines ANP and BANP policies to isolate egress traffic to a range of specified CIDR ranges.
```

kind: AdminNetworkPolicy
metadata:
  name: network-as-egress-peer
spec:
  priority: 70
  subject:
    namespaces: {}
  egress:
  - name: "deny-egress-to-external-dns-servers"
    action: "Deny"
    to:
    - networks:
      - 8.8.8.8/32
      - 8.8.4.4/32
      - 208.67.222.222/32

>     - pods:

Reviewer 3: this means you have denied egress traffic to also all your pods in your cluster right?, the rule name is not saying that... please help me understand why do we have this part

        namespaceSelector: {}
        podSelector: {}

>   - name: "allow-all-egress-to-intranet"

Reviewer 3: intranet cannot be 1 IP make this /24 so that it matches the intent, and please change it from 172.30 which is the service CIDR usually in OCP

Author: updated

    action: "Allow"
    to:
    - networks:
      - 172.30.0.10/32 <1>

>     - pods:

Reviewer 3: umm on rule1 you are denying to all pods on your cluster, on rule2 we are allowing to all pods on your cluster and on rule3 we are again allowing to all pods on your cluster, maybe there was a copy paste snafu somewhere?

        namespaceSelector: {}
        podSelector: {}
    ports:
      - portNumber:
          protocol: UDP

>           port: 53

Reviewer 3: this is supposed to be in the DNS deny rule right? 53 is DNS port

>   - name: "allow-all-intra-cluster-traffic"

Reviewer 3: https://github.com/openshift/openshift-docs/pull/75886#discussion_r1639599314 this comment has not being addressed yet so the issue still stands

      - 203.0.113.0/24
      - 198.51.100.0/24
  - name: "allow-all-intra-cluster-traffic"
    action: "Allow"
    to:
    - networks: <2>
      - <POD_CIDR>
      - <NODE_CIDR>
      - <SERVICE_CIDR>
  - name: "pass-all-egress-to-internet"
    action: "Pass"
    to:
    - networks:
      - 0.0.0.0/0
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces: {}
  egress:
  - name: "deny-all-egress-to-internet"
    action: "Deny"
    to:
    - networks: <3>
      - 0.0.0.0/0
----
<1> Use `networks` to specify a range of CIDR networks outside of the cluster.

> <2> Specifies the ranges for the intra cluster traffic from the your <POD_CIDR>, <NODE_CIDR>, <SERVICE_CIDR> resources.

Reviewer 4: ```suggestion
<2> Specifies the ranges for the intra-cluster traffic from the <POD_CIDR>, <NODE_CIDR>, and <SERVICE_CIDR> resources.
```
Took a shot at rewording this

Author: Had to change this example.

> <3> Specifies a strong `Deny` egress policy to the internet by setting `networks` values to - 0.0.0.0/0. Make sure you have an allow policy at a higher priority before setting a `Deny` to 0.0.0.0/0 because this will deny all traffic including to API servers.

Reviewer 3: I'd make this a red like with EFW but I leave it to you, 
```suggestion
<3> Specifies a `Deny` egress  to everything by setting `networks` values to - 0.0.0.0/0. Make sure you have higher priority allow rules to necessary destinations before setting a `Deny` to 0.0.0.0/0 because this will deny all traffic including to Kubernetes API and DNS servers.
```

====

> Collectively the "Example of ANP and BANP using `networks` peers enforces the following egress policy:

Reviewer 3: ```suggestion
Collectively the `network-as-egress-peer` ANP and `default` BANP using `networks` peers enforces the following egress policy:
```

> * All pods cannot talk to company's intranet DNS servers.

Reviewer 3: 8.8.8.8 is not company's intranet DNS

Reviewer 3: ^ you forgot this, now that we have edited the policy change these 4 points please,

Author: All pods cannot talk to internet DNS servers. Is this the correct interpretation of the policy? @tssurya

* All pods can talk to rest of the company's intranet.

> * All pods can talk to other pods, nodes, services.

Reviewer 4: ```suggestion
* All pods can talk to other pods, nodes, and services.
```

> * All pods cannot talk to internet. Combining the last ANP `Pass` rule and the strong BANP `Deny` a guardrail policy is created that secures traffic in the cluster.

Reviewer 4: ```suggestion
* All pods cannot talk to the internet. Combining the last ANP `Pass` rule and the strong BANP `Deny` rule, a guardrail policy is created that secures traffic in the cluster.
```

[id="combined-nodes-peer-networks-peer-anp"]
== Using nodes peer and networks peer together

> Cluster Administrators can combine `nodes` and `networks` peer in your ANP and BANP policies.

Reviewer 4: ```suggestion
Cluster administrators can combine `nodes` and `networks` peers in ANP and BANP policies.
```

.Example of combined `nodes` and `networks` peer
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: egress-peer-1 <1>
spec:
  egress: <2>
  - action: "Allow"
    name: "allow-egress"
    to:
    - nodes: <3>
        matchExpressions:
        - key: worker-group
          operator: In
          values:
          - workloads # Egress traffic from nodes with label worker-group: workloads is allowed.

>     - networks:

Reviewer 3: 4 annotation for networks is missing?

Author: Sorry, I'm not sure what happened here. I mislabeled. I fixed these. I didn't think it was worth re-annotating this example too heavily given the comments so I tried to emphasize higher level things `name` priorit, etc

            apps: external-apps
        podSelector:
          matchLabels:
            app: web # This rule in the policy allows the traffic directed to pods labeled apps: web in projects with apps: external-apps to leave the cluster.
  - action: "Deny"
    name: "deny-egress"
    to:
    - nodes:
        matchExpressions:
        - key: worker-group
          operator: In
          values:
          - infra # Egress traffic from nodes with label worker-group: infra is denied.
    - networks:
      - 104.154.164.160/32 # Egress traffic to this IP address from cluster is denied.
    - pods:
        namespaceSelector:
          matchLabels:
            apps: internal-apps
        podSelector: {}
  - action: "Pass"
    name: "pass-egress"
    to:
    - nodes:
        matchExpressions:
        - key: node-role.kubernetes.io/worker
          operator: Exists # All other egress traffic is passed to NetworkPolicy or BANP for evaluation.
  priority: 30 <3>
  subject: <4>
    namespaces:
      matchLabels:
        apps: all-apps
----
<1> Specifies the name of the policy.
<2> For `nodes` and `networks` peers, you can only use northbound traffic controls in ANP as `egress`.

> <3> Specifies the priority of the ANP, determining the order in which they should be evaluated. Lower priority rules have higher precedence. ANP accepts values of 0-99 with 0 being the highest priority and 99 being the lowest.

Reviewer 3: I see two `3's` (by chance :D I hope peer reviewing is also done because I am not very familiar with the annotation bits in docs)

Author: I got all mixed up in my numbering when we updated the policy. I fixed it now.

> <3> Specifies the priority of the ANP determines the order in which they should evaluated. The lower the number the higher precedence. ANP accepts 100 values of 0-99 with 0 being the highest priority and 99 being the lowest.

Reviewer 4: ```suggestion
<3> Specifies the priority of the ANP, determining the order in which they should be evaluated. Lower priority rules have higher precedence. ANP accepts values of 0-99 with 0 being the highest priority and 99 being the lowest.
```
Reworded for clarity

Author: good rewording, thank you.

> <4> Specifies the set of pods in cluster on which the rules of the policy are to be applied. In the example, any pods with `apps: all-apps` across all namespaces are the `subject` of the policy.

Reviewer 4: ```suggestion
<4> Specifies the set of pods in the cluster on which the rules of the policy are to be applied. In the example, any pods with the `apps: all-apps` label across all namespaces are the `subject` of the policy.
```

* Implement egress traffic control towards Kubernetes API servers using `nodes` or `networks` egress peers

* Implement egress traffic control towards external destinations outside the cluster using `networks` peer

[id="egress-traffic-control-towards-cluster-nodes_{context}"]
== Using nodes peer to control egress traffic to cluster nodes

Using the `nodes` peer administrators can control egress traffic from pods to nodes in the cluster.

The following example allows egress traffic to the Kubernetes API server on port `6443` by any of the namespaces with a `restricted`, `confidential`, or `internal` level of security and in the `dept: engr` tenancy.

.Example of ANP `Allow` egress using `nodes` peer
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: egress-security-allow
spec:
  egress:
  - action: Deny
    to:
    - nodes:
        matchExpressions:
        - key: node-role.kubernetes.io/worker
          operator: Exists
  - action: Allow
    name: allow-to-kubernetes-api-server
    ports:
    - portNumber:
        port: 6443
        protocol: TCP

>     - portNumber:

Reviewer: @JoeAldinger We can get rid of this block of ports in policy as we are pointing to port 6443 only. The port on which endpoints of kube api server listens.

        port: 443
        protocol: TCP
    to:
    - nodes:
        matchExpressions:
        - key: node-role.kubernetes.io/control-plane <1>
          operator: Exists
    - pods:
        namespaceSelector:
          matchLabels:
            dept: engr <2>
        podSelector: {}
  priority: 55
  subject: <3>
    namespaces:
      matchExpressions:
      - key: security
        operator: In
        values: <4>
        - restricted
        - confidential
        - internal
----
<1> Specifies a node or set of nodes in the cluster using the `matchExpressions` field. It will select all nodes if the label is present, but left empty.
<2> Specifies all the pods labeled with `dept: engr`.

> <3> Specifies the subject of the ANP which includes any namespaces that match the labels used by the network policy. The example matches any of the namespaces with `restricted`, `confidential`, or `internal` level of security.

Reviewer: Could include security also in single quotes to high light key in example.

Author: I can but the reason it isn't in `security` is because I mean it in a general sense not the literal.

<4> Specifies key/value pairs for `matchExpressions` field.
====

[id="egress-traffic-control-networks-peer-external-destinations_{context}"]
== Using networks peer to control egress traffic towards external destinations

> Cluster administrators can also use CIDR ranges for `networks` peer and apply an egress traffic policy to those nodes. The following are benefits to using `nodes` and `networks` peer when blocking access from cluster workloads to specific ports on cluster nodes:

Reviewer: IMO The content from The following are benefits.. and three bullet items should be on the top as they are applicable to both nodes and networks peers not specific to network peers. After line number 16. 

Last two bullets are for nodes peers

Reviewer: Alternatively move them under last section of using nodes and networks both.

Author: I moved the one up to the bullet points where you wanted it. The other I used as a framing sentence for this section and the third one I dropped because there was outstanding debate on it.

.Example of ANP and BANP using `networks` peers
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: network-as-egress-peer
spec:
  priority: 70
  subject:
    namespaces: {}
  egress:
  - name: "deny-egress-to-external-dns-servers"
    action: "Deny"
    to:
    - networks: <1>
      - 8.8.8.8/32
      - 8.8.4.4/32
      - 208.67.222.222/32
  - name: "allow-all-egress-to-intranet"
    action: "Allow"
    to:
    - networks:
      - 172.30.0.10/32
    ports:
      - portNumber:
          protocol: UDP
          port: 53
  - name: "allow-all-intra-cluster-traffic"
    action: "Allow"
    to:
    - networks:
      - 10.128.0.0/14
      - 172.30.0.0/16 <2>
  - name: "pass-all-egress-to-internet"
    action: "Pass"
    to:
    - networks:
      - 0.0.0.0/0 <3>
---
apiVersion: policy.networking.k8s.io/v1alpha1
kind: BaselineAdminNetworkPolicy
metadata:
  name: default
spec:
  subject:
    namespaces: {}
  egress:
  - name: "deny-all-egress-to-internet"
    action: "Deny"
    to:
    - networks:
      - 0.0.0.0/0 <3>
---

> apiVersion: policy.networking.k8s.io/v1alpha1

Reviewer: The BANP is included twice in example YAML
